### ISSUE TITLE
[OpenMP] runtime might try to take the bootstrap lock twice and deadlock

### ISSUE THREAD

**Issue by jprotze** [llvm/llvm-project#86684]
This was initially reported here (including stacktraces): https://stackoverflow.com/questions/78183545/does-compiling-imagick-with-openmp-enabled-in-freebsd-13-2-cause-sched-yield

If `__kmp_register_library_startup` detects that another instance of the library is present, `__kmp_is_address_mapped` is eventually called. It calles `__kmp_str_format`, which uses `kmpc_alloc` to allocate memory for the string. This function calls `__kmp_entry_thread` to access the thread-local memory pool, which seems a bad idea during initialization. This macro internally calls `__kmp_get_global_thread_id_reg` which set the bootstrap look at the beginning (before calling `__kmp_register_library_startup`).

**Comment by caliguian** [comment]
It would be fantastic if someone (more knowledgeable than I) could look into this. We've been manually killing the stuck processes every day for quite some time now, which is not a good time. The processes get locked with an infinite sched_yield() call that maxes out the CPU until we see it and manually kill it.

**Comment by jprotze** [comment]
@jpeyton52 can you look into this issue?

**Comment by jpeyton52** [comment]
@caliguian, it may help to manually delete `/tmp/__KMP_REGISTERED_LIB_*` files as well if they exist. This is especially true if you begin seeing a message like the following after using the fixed library (from PR #88539):

```
This means that multiple copies of the OpenMP runtime have been linked into the program.
That is dangerous, since it can degrade performance or cause incorrect results. The best thing
to do is to ensure that only a single OpenMP runtime is linked into the process, e.g. by avoiding
static linking of the OpenMP runtime in any library. As an unsafe, unsupported, undocumented
workaround you can set the environment variable KMP_DUPLICATE_LIB_OK=TRUE to allow 
the program to continue to execute, but that may cause crashes or silently produce incorrect results. 
```

**Comment by caliguian** [comment]
@jpeyton52 Thank you!
